src/sftp/mod.rs

```diff
@@ -0,0 +1,309 @@
+            read_task,
+            flush_task,
+        })
+    }
+
+    /// Close sftp connection
+    ///
+    /// # Cancel Safety
+    ///
+    /// This function is cancel safe.
+    pub async fn close(self) -> Result<(), Error> {
+        // Flush the data.
+        //
+        // Since there is no reference to `Sftp`, the only requests that
+        // haven't yet flushed should be close requests.
+        //
+        // And there will not be any new requests.
+        flush(&self.shared_data).await?;
+
+        // Terminate flush_task, otherwise read_task would not return.
+        self.flush_task.abort();
+        match self.flush_task.await {
+            Ok(res) => res?,
+            Err(join_err) => {
+                if !join_err.is_cancelled() {
+                    return Err(join_err.into());
+                }
+            }
+        }
+
+        // Drop the shared_data, otherwise read_task would not return.
+        debug_assert_eq!(self.shared_data.strong_count(), 2);
+        drop(self.shared_data);
+
+        // Wait for responses for all requests buffered and sent.
+        self.read_task.await??;
+
+        let res: Result<ExitStatus, Error> =
+            crate::child::delegate!(self.child, child, { child.wait().await });
+        let exit_status = res?;
+
+        if !exit_status.success() {
+            Err(Error::SftpError(
+                openssh_sftp_client::Error::SftpServerFailure(exit_status),
+            ))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn write_end(&self) -> WriteEndWithCachedId {
+        WriteEnd::new(self.shared_data.clone()).into()
+    }
+
+    /// Get maximum amount of bytes that one single write requests
+    /// can write.
+    pub fn max_write_len(&self) -> u32 {
+        self.shared_data.get_auxiliary().limits().write_len
+    }
+
+    /// Get maximum amount of bytes that one single read requests
+    /// can read.
+    pub fn max_read_len(&self) -> u32 {
+        self.shared_data.get_auxiliary().limits().read_len
+    }
+
+    /// Return a new [`OpenOptions`] object.
+    pub fn options(&self) -> OpenOptions<'_> {
+        OpenOptions::new(self)
+    }
+
+    /// # Cancel Safety
+    ///
+    /// This function is cancel safe.
+    pub async fn create(&self, path: impl AsRef<Path>) -> Result<File<'_>, Error> {
+        self.options()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(path)
+            .await
+    }
+
+    /// # Cancel Safety
+    ///
+    /// This function is cancel safe.
+    pub async fn open(&self, path: impl AsRef<Path>) -> Result<File<'_>, Error> {
+        self.options().read(true).open(path).await
+    }
+
+    /// * `cwd` - The current working dir for the [`Fs`].
+    ///           If `cwd` is empty, then it is set to use
+    ///           the default directory set by the remote
+    ///           `sftp-server`.
+    pub fn fs(&self, cwd: impl Into<PathBuf>) -> Fs<'_> {
```

Comment:
@jonhoo I need some help here.

I somehow mess up the lifetime for `Fs` here.

During testing, I was able to call `Sftp::close`, which takes `Sftp` by value, without causing compilation error due to existence of `Fs`.

Reply:
Somehow, `sftp::File` does not suffer from this problem.

Holding `sftp::File` when `sftp::Sftp` is `close`d would cause a compile time error.

This is quite confusing for me, since `sftp::File` and `sftp::Fs` has the same lifetime annotations.

Reply:
This bug is fixed after I introduce lifetime to `WriteEndWithCachedId`.